src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -215,15 +215,26 @@ public override void GetObjectData(SerializationInfo info, StreamingContext cont
         /// </summary>
         public Dispatcher Dispatcher { get; }
 
-        public IList<SiloAddress> GetCompatibleSiloList(GrainId grain)
+        public IList<SiloAddress> GetCompatibleSiloList(PlacementTarget target)
         {
             // For test only: if we have silos that are not yet in the Cluster TypeMap, we assume that they are compatible
             // with the current silo
             if (this.config.AssumeHomogenousSilosForTesting)
                 return AllActiveSilos;
 
-            var typeCode = grain.GetTypeCode();
-            var compatibleSilos = GrainTypeManager.GetSupportedSilos(typeCode).Intersect(AllActiveSilos).ToList();
+            var typeCode = target.GrainId.GetTypeCode();
+            IReadOnlyList<SiloAddress> silos;
+            if (target.InterfaceVersion > 0)
+            {
+                var versions = new[] { GrainTypeManager.GetAvailableVersions(target.InterfaceId).Max() };
+                silos = GrainTypeManager.GetSupportedSilos(typeCode, target.InterfaceId, versions);
```

Comment:
IMO, the default safe choice would be to use the version that was requested, not the maximum version